[ES|QL] Use custom AST for suggestion, validation, etc...

dej611 opened the pull request:
## Summary

Fixes #166242
This is a PR to use an higher level AST instead of the raw ParserTree to compute all Monaco features for ES|QL.

## Grammar

This PR drops the existing grammar in favour of the ES one, which a tiny tweak in order to support any case for the query.
I think we could fully adopt the ES grammar without tweaks once [this issue](https://github.com/tunnelvisionlabs/antlr4ts/issues/535) is resolved in the antlr4ts dependency.

The adoption of the ES grammar helped also to resolve #166173 .

<img width="407" alt="Screenshot 2023-09-21 at 12 16 45" src="https://github.com/elastic/kibana/assets/924948/38aa507b-8371-4387-ab8a-95423cdab684">
<img width="427" alt="Screenshot 2023-09-21 at 12 16 40" src="https://github.com/elastic/kibana/assets/924948/462d9b57-24cf-45f8-bcf1-070bbdb67817">
<img width="444" alt="Screenshot 2023-09-21 at 12 16 29" src="https://github.com/elastic/kibana/assets/924948/a6b5a904-c16f-40c6-a377-4e3e2005975f">
<img width="415" alt="Screenshot 2023-09-21 at 12 16 23" src="https://github.com/elastic/kibana/assets/924948/e7732f91-17b2-4af5-9d52-0eb96fa1999e">
<img width="404" alt="Screenshot 2023-09-21 at 12 16 17" src="https://github.com/elastic/kibana/assets/924948/99311f65-1390-4a90-8bb6-12b263d808fd">

## AST generator

The AST is a subset of the ESQL grammar parser tree, which is easier to navigate.
Tests will be provided to the tool in order to test that the tool can translate even the craziest query string.
**Note** that this is not preventing the query to be submitted to ES: even if a validation case fails here or the AST fails to translate part of the expression ES will always have the last word about error and warnings, which will be promptly notified to the user.

Commands coverage:
* `ROW`
  * [x] `ROW a=1`
  * [x] `ROW a=1, b=1`
* `FROM`
  * [x] `FROM a`
  * [x] `FROM a, b`
  * [x] `FROM a [METADATA ...]`
  * [x] `FROM a, b [METADATA ...]` 
    * `METADATA` won't have indentifiers yet
* `SHOW`
  * [x] `INFO`
  * [x] `FUNCTIONS`
* `EVAL`
  * [x] `EVAL a = b + 1`
  * [x] `EVAL a = fnA( ... ) + fnB( ... )`
  * [x] `EVAL b + 1`
  * [x] `EVAL fnA( ... ) + fnB( ... )`
* `STATS`
  *  [x] `STATS a = fn(b) + fn(c)`
  *  [x] `STATS a = fn(b) + fn(c) BY group`
  *  [x] `STATS a = fn(b) + fn(c) BY groupA, groupB`
  *  [x] `STATS fn(b) + fn(c)`
  *  [x] `STATS fn(b) + fn(c) BY groupA`
  *  [x] `STATS fn(b) + fn(c) BY groupA, groupB`
* [x] `WHERE`
* [x] `LIMIT` 
* [x] `KEEP`
* [x] `PROJECT`
* [x] `DROP`
* `RENAME`. 
  * [x] `RENAME a as b`  
  * [x] `RENAME `a + 1 / 5` as `myField`
* [x] `DISSECT`
* [x] `GROK`
* [x] `ENRICH`
* [x] `MV_EXPAND`
* [x] `SORT`

Collect all user defined variables
  * [x] Basic support for assigned variables (i.e. `eval a = ...`)
  * [x] Field type shadowing with assignment
  * [x] Advanced support for non-assigned cases? (i.e. `eval a + 1`)

## Validation

With this new approach it would be possible to catch most of warnings and errors while typing, preventing users to submit invalid queries to ES.
Note that here the validation function can detect multiple errors and warnings at once, while ES only returns one error at the time - improving the current sub-optimal editing experience provided.

![esql_clientside_validation](https://github.com/elastic/kibana/assets/924948/c02f057c-efb9-40b6-ad4a-9642a11c8367)

Tasks
* Create definitions
  * [x] Define function definitions
  * [x] Define aggs definitions
  * [x] Define command definitions
  * [x] Define command options definitions
* Create basic validation walker function
  * [x] Support both errors and warnings
  * [x] Validate commands
  * [x] Validate command options
  * [x] Validate functions
  * [x] Validate source/columns (I phase - just formally)
  * [x] Validate policies (I phase - just formally)
  * [x] Validate literals
  * [x] Validate duration/math syntax
  * [x] Validate variables
* Improve validation work with high level checks
  * [x] Validate source/columns (II phase - leverage ES metadata)
  * [x] Validate policies (II phase - leverage ES metadata)
  * [x] Optimize queries
    * phase I: just make queries strictly when required
* Write tests
  * [x] Source commands
    * [x] Options
  * [x] Processing commands
    * [x] Options
  * [x] Sorting commands
    * [x] All options, combinations
  * [x] Enrich commands
    * [x] Options
  * [x] Warnings cases
* [x] Integrate all of it as Monaco diagnostics module
* [x] Solve issue with long queries and single line mode
* Provide better sync between client and server side error management
  * [x] Show server side warnings only on absence of errors
  * [x] Clear up server side errors on code change

## Autocomplete

Using the same provider here to provide better contextual suggestions.
First example giving better suggestions (filtered field based on types and functions based on returned type):

![better_suggestions_types](https://github.com/elastic/kibana/assets/924948/bffe158f-bb31-47cf-9cd6-8bdcade41409)

* Commands
  * [x] Show all commands on empty string
  * [x] Show all but source commands after first pipe
  * [x] Suggest option at the right moment

## Hover

Basic hover information about used functions:
<img width="688" alt="Screenshot 2023-10-16 at 16 22 16" src="https://github.com/elastic/kibana/assets/924948/e623c71d-4bbc-4511-8bb8-d818e465e554">
<img width="533" alt="Screenshot 2023-10-16 at 16 21 54" src="https://github.com/elastic/kibana/assets/924948/af7a244f-be5b-46a0-8704-09c1ffc48a9e">

# Things not covered in this PR

* Load function definitions from `SHOW FUNCTIONS` at language bootstrap
* Make `validate` function an independent module?
  * Provide a "formal" validation option (i.e. mute `Unknown <entity> [...]` messages)
* Have `autocomplete` on par with the current state in `main`

## Validation cases not covered yet

* `... | eval var = 1 year` => ❌ it should report an error
* `... | eval 1 + [1, 2, 3]` => ⚠️ it should report a null warning
  * same applies for `-`, `*`, `/`, `%`
* any location aware field/variable check
  * `... | stats a = avg(field) | keep b` => ❌ `b` should be reported as unknown
  * `... | keep a | eval b` => ❌  `b` should be reported as unknown
  * `... | rename a as b, d as c, c as b` => ❌ `c` should be reported as unknown
  * `... | enrich policy on fieldFromEnrich` => ❌ `fieldFromEnrich` is not reported as unknown
  * `... | eval a = round(b), b = round(a)` => ❌  `b` should be reported an unknown
* `from ... [metadata <list of meta fields>` => ❌  `meta fields` won't be validated as there's currently no available API to get them

## Hover

Advanced usage of hover feature.

## Signature

TBD

dej611 commented:
Not sure what happened and why the `observability` change landed here 🤔

stratoula commented:
+460 KB on the sync bundle doesn't good Marco. Can we improve?

dej611 commented:
> +460 KB on the sync bundle doesn't good Marco. Can we improve?

I'm trying to figure out where that number comes from, but it's on the shared bundle so it is not inspectable.
Anyway now that all tests are passing I'm on it.

stratoula reviewed (commented):
Some notes from me (I haven't checked the code yet)

- For some reason I get a 0 warning icon
<img width="2497" alt="image" src="https://github.com/elastic/kibana/assets/17003240/f345f80b-e9d9-49f2-b564-1fcd8b6a4d8a">

This works ok in main ( I should see 1 warning and also the warning on the footer but I don't)

- I see the error for unknown index but is wrong (I have multiple logstash indices and I want to use the wildcard)
<img width="876" alt="image" src="https://github.com/elastic/kibana/assets/17003240/4a4f4e50-31ab-4607-a96f-e5790618902c">

- I think this is important, I want to see the list of aggregations here. 

<img width="982" alt="image" src="https://github.com/elastic/kibana/assets/17003240/950773e2-2745-4923-91a9-dca360aaf9a9">

- Here also 
<img width="990" alt="image" src="https://github.com/elastic/kibana/assets/17003240/bbeb843b-ead5-4c10-aedb-f0db53dde07c">

If we want to work on autocomplete on a follow up PR can we merge this in a feature branch instead and not main. I feel that the regressions in autocomplete are important and we should not  merge them in main.

- Seems the validation hasn't run properly here

<img width="988" alt="image" src="https://github.com/elastic/kibana/assets/17003240/02720ed5-99f9-4df3-a10b-e263ae31df6a">

- Here I expect var1 to be in the list of suggestions as it is on stats

<img width="1193" alt="image" src="https://github.com/elastic/kibana/assets/17003240/7f1c1eec-3564-4c30-8785-46c4f91d65a7">

dej611 commented:
Thanks for your quick feedback @stratoula .
I'd prefer to split validation or hover issue from autocomplete as the latter deserve a more focused approach and there's already a lot of content here. I'm ok to have a feature branch to merge in before `main` and work on a separate autocomplete PR as follow up.

Will reply inline here about validation and UI issues, will defer the autocomplete ones for a later PR:

> * For some reason I get a 0 warning icon
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/17003240/276164684-f345f80b-e9d9-49f2-b564-1fcd8b6a4d8a.png">
> This works ok in main ( I should see 1 warning and also the warning on the editor footer but I don't)

Can reproduce as well, fixed with https://github.com/elastic/kibana/pull/166185/commits/9acd4361c8480841ba3e723dd733650b97b76d87

> 
> * I see the error for unknown index but is wrong (I have multiple logstash indices and I want to use the wildcard)
> 
> <img alt="image" width="876" src="https://user-images.githubusercontent.com/17003240/276165547-4a4f4e50-31ab-4607-a96f-e5790618902c.png">

This is a bug as I was not validating fuzzy strings for indices. Added the check and a unit test for it with https://github.com/elastic/kibana/pull/166185/commits/8ec22264a5b10f5232393b7739d7842eab27d2f4

> * Seems the validation hasn't run properly here
> 
> <img alt="image" width="988" src="https://user-images.githubusercontent.com/17003240/276167652-02720ed5-99f9-4df3-a10b-e263ae31df6a.png">

Can reproduce as well, fixed with https://github.com/elastic/kibana/pull/166185/commits/9acd4361c8480841ba3e723dd733650b97b76d87

dej611 commented:
It seems that this PR is actually reducing the bundlesize of the `kbn-monaco` bundle rather than increase it.

Enabling detailed stats on the `kbn-monaco` webpack optimizer I get this numbers:
```
[ibazel]     Hash: 8f992c2e72951cec4d98
[ibazel]     Time: 1797ms
[ibazel]     Built at: 10/18/2023 12:43:36 PM
[ibazel]                         Asset      Size  Chunks                   Chunk Names
[ibazel]         esql.editor.worker.js  1.75 MiB    main  [emitted]        main
[ibazel]     esql.editor.worker.js.map  1.85 MiB    main  [emitted] [dev]  main
[ibazel]     Entrypoint main = esql.editor.worker.js esql.editor.worker.js.map
[ibazel]     chunk {main} esql.editor.worker.js, esql.editor.worker.js.map (main) 1.56 MiB [entry] [rendered]
```

While on `main` I get this:
```
[ibazel]     Hash: 2b4a5e316705efa60531
[ibazel]     Time: 2294ms
[ibazel]     Built at: 10/18/2023 12:45:28 PM
[ibazel]                         Asset      Size  Chunks                   Chunk Names
[ibazel]         esql.editor.worker.js  2.04 MiB    main  [emitted]        main
[ibazel]     esql.editor.worker.js.map  2.15 MiB    main  [emitted] [dev]  main
[ibazel]     Entrypoint main = esql.editor.worker.js esql.editor.worker.js.map
[ibazel]     chunk {main} esql.editor.worker.js, esql.editor.worker.js.map (main) 1.79 MiB [entry] [rendered]
```

so `kbn-monaco` get ~200kb lighter from the stats.

After looking at the shared-bundle-src detailed view maybe I get what's going on here:

```diff
[ibazel]                                 Asset       Size                  Chunks                          Chunk Names
[ibazel]  a1cef3d530e1adb0f52b2f62994a2aca.ttf  69.3 KiB                          [emitted]       
- [ibazel]     kbn-ui-shared-deps-src.chunk.0.js   1.37 MiB                       0  [emitted]        [big]  
+ [ibazel]     kbn-ui-shared-deps-src.chunk.0.js   466 KiB                       0  [emitted]        [big]  
- [ibazel] kbn-ui-shared-deps-src.chunk.0.js.map   1.42 MiB                       0  [emitted] [dev]   
+ [ibazel] kbn-ui-shared-deps-src.chunk.0.js.map   438 KiB                       0  [emitted] [dev]              
- [ibazel]     kbn-ui-shared-deps-src.chunk.1.js    466 KiB                       1  [emitted]        [big]     
+ [ibazel]     kbn-ui-shared-deps-src.chunk.1.js  14.6 KiB                       1  [emitted]             
- [ibazel] kbn-ui-shared-deps-src.chunk.1.js.map    438 KiB                       1  [emitted] [dev]     
+ [ibazel] kbn-ui-shared-deps-src.chunk.1.js.map  14.2 KiB                       1  [emitted] [dev]         
- [ibazel]     kbn-ui-shared-deps-src.chunk.2.js   14.6 KiB                       2  [emitted]            
+ [ibazel]     kbn-ui-shared-deps-src.chunk.2.js  4.94 KiB                       2  [emitted]                 
- [ibazel] kbn-ui-shared-deps-src.chunk.2.js.map   14.2 KiB                       2  [emitted] [dev]     
+ [ibazel] kbn-ui-shared-deps-src.chunk.2.js.map  4.35 KiB                       2  [emitted] [dev]       
- [ibazel]     kbn-ui-shared-deps-src.chunk.3.js   1.33 KiB                       3  [emitted]                    
+ [ibazel]     kbn-ui-shared-deps-src.chunk.3.js  7.61 KiB                       3  [emitted]               
- [ibazel] kbn-ui-shared-deps-src.chunk.3.js.map  124 bytes                       3  [emitted] [dev]       
+ [ibazel] kbn-ui-shared-deps-src.chunk.3.js.map  7.87 KiB                       3  [emitted] [dev]        
- [ibazel]     kbn-ui-shared-deps-src.chunk.4.js   4.94 KiB                       4  [emitted]               
- [ibazel] kbn-ui-shared-deps-src.chunk.4.js.map   4.35 KiB                       4  [emitted] [dev]         
- [ibazel]     kbn-ui-shared-deps-src.chunk.5.js   7.61 KiB                       5  [emitted]               
- [ibazel] kbn-ui-shared-deps-src.chunk.5.js.map   7.87 KiB                       5  [emitted] [dev]         
[ibazel]            kbn-ui-shared-deps-src.css   92.2 KiB  kbn-ui-shared-deps-src  [emitted]               kbn-ui-shared-deps-src
[ibazel]        kbn-ui-shared-deps-src.css.map    120 KiB  kbn-ui-shared-deps-src  [emitted] [dev]         kbn-ui-shared-deps-src
- [ibazel]             kbn-ui-shared-deps-src.js   7.47 MiB  kbn-ui-shared-deps-src  [emitted]        [big]  kbn-ui-shared-deps-src
+ [ibazel]             kbn-ui-shared-deps-src.js  8.97 MiB  kbn-ui-shared-deps-src  [emitted]        [big]  kbn-ui-shared-deps-src
- [ibazel]         kbn-ui-shared-deps-src.js.map    6.5 MiB  kbn-ui-shared-deps-src  [emitted] [dev]         kbn-ui-shared-deps-src
+ [ibazel]         kbn-ui-shared-deps-src.js.map  8.01 MiB  kbn-ui-shared-deps-src  [emitted] [dev]         kbn-ui-shared-deps-src
```

Apparently part of the `chunk.0` and `chunk.3` in this PR gets incorporated into the `kbn-ui-shared-deps-src.js` initial bundle. I'll investigate whether it's possible to move it back to an async bundle the `kbn-monaco` package, but I wonder if its actual shrink may have played some role in here.

The good news is that overall this PR adds a whopping *12.8 Kb* overall to the shared bundle.
The bad news is that this increase seems to be shifted on the wrong side.

dej611 commented:
Managed to have a view of what is going on here and why the bundle changed.

This PR moved both `esql` and `antlr4ts` into the main bundle:

<img width="1501" alt="Screenshot 2023-10-19 at 09 24 11" src="https://github.com/elastic/kibana/assets/924948/e6477e06-79fd-4065-bd65-d341b1451bb3">

While in `main` both modules are in the async chunk (not tracked by the github metrics):

<img width="1505" alt="Screenshot 2023-10-19 at 09 24 02" src="https://github.com/elastic/kibana/assets/924948/9d786fa9-bf42-43b4-8561-c18f9da26ed9">

I'll try to move back both modules back to the async realm.

dej611 commented:
Managed to solve the bundle issue and will report here some things I found during the investigation which might be useful if somebody is facing the same problem with the shared bundle:

* Mind your `index.ts` "barrel files"
  * this is not something strictly a performance problem as described [in this article](https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7), but rather a dependency graph problem. The tree-shaking power of a bundler is limited and adding more depth in the analysis between two modules with middle men like "barrel files" might make it harder to optimized (for the bundler) and debug (for the developer) this type of problems.

* if you can't see it, it's hard to debug
  * it's really hard to debug something you do not see, especially when you have to track down dependencies and where a package is to determine the final bundle size. Right now it is not easy to analyse bundle size of something that is not a plugin, as well documented [in the Kibana documentation](https://www.elastic.co/guide/en/kibana/current/plugin-performance.html#_understanding_plugin_bundle_size). So for this specific use case it was required to manual tweak the Bazel + Webpack configuration in order to emit a `stats.json` file that can be visualized.
  * Tools to visualize here are fundamental to understand the relationships of various packages, and the combination of several representation is key to find the problem: [`Webpack bundle analyzer`](https://www.npmjs.com/package/webpack-bundle-analyzer) was useful to understand in which chunk the package `kbn-monaco`, `antlr4ts` and more were stored, while the [`Webpack analyse > Modules`](https://webpack.github.io/analyse) app was super useful to track the type of `imports` (async vs sync) between files.

* Mind static and class-based properties. This code got me some headache to refactor in a way that was treeshaking properly:

```ts
const symbolsLookup: Record<number, string> = Object.entries(esql_parser)
  .filter(([k, v]) => typeof v === 'number' && !/RULE_/.test(k) && k.toUpperCase() === k)
  .reduce((memo, [k, v]: [string, number]) => {
    memo[v] = k;
    return memo;
  }, {} as Record<number, string>);
```

The problem here was that the compiler didn't know what to keep and what to drop as the code was iterating thru everything there. The code was needed to map rule numbers to their string equivalent, but luckily there's a static property to use which makes the tree-shake work performantly:

```ts
esql_parser.VOCABULARY.getSymbolicName(<tokenId>);
```
* if you have a big dependency, moving the logic who depends on that on a Web Worker moves the big dependency on an async chunk.
  * in this case generating the simplified AST is not much code (`ast_factory.ts` + `ast_walker.ts`) per se, but it requires the `esql_parser` full dependency, which in turns requires the full `antlr4ts` library, which is worth 1 Mb. Moving the AST generation on the web worker makes the architecture and the flow a little bit more convoluted (see the new `packages/kbn-monaco/src/esql/lib/monaco/esql_ast_provider.ts` adapter) but it ensures that the 1 Mb dependency is required only within the web worker as async chunk.

dej611 commented:
As discussed offline with @stratoula this PR now targets the new feature branch `feature/esql-validation` where next iterations with autocomplete and fetching optimizations will land.

Once this PR will get reviewed (validation & hover) and merged into the new feature branch the following follow ups will be pursued:
* fix autocomplete
  * in particular put `stats` and `eval` on par with current main state
* fix number of requests for validation
  * not allow it to make requests actively and only enable field validation when the cache gets populated by autocomplete
  * same for hover

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

davismcphee commented on test/functional/apps/discover/group3/_request_counts.ts:
```diff
@@ -223,7 +223,8 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
     });
 
-    describe('ES|QL mode', () => {
+    // @TODO: fix this in a follow up
```
I didn't realize this was going into a feature branch at first and was surprised to see these tests skipped, but assuming they'll be addressed and unskipped before making it to `main`, this LGTM 👍

stratoula replied:
Davis correct, we found some performance problems so this is not going to be merged. Also autocomplete doesnt work as expected. But as this is a big PR already we want to move this into a feature branch and then continue with the investigation and the development

stratoula reviewed (approved):
As discussed with Marco this is not ready to be merged this is why we decided to create a feature branch instead. Specifically:

- autocomplete needs some refinement
- We should see what we are going to about making the client side validation performant. Currently we are hitting ES every 250ms and also we add another api call to Discover initialization

 As a first step the code looks ok to me so I am approving the merge to the feature branch

Dosant reviewed (approved):
I think it would be great if you'd make your team the owner of kbn-monaco/src/esql :D
